plotter/contour.go

```diff
@@ -0,0 +1,488 @@
+
+//go:generate ./list
+
+package plotter
+
+import (
+	"image/color"
+	"math"
+	"sort"
+
+	"github.com/gonum/plot"
+	"github.com/gonum/plot/palette"
+	"github.com/gonum/plot/vg"
+	"github.com/gonum/plot/vg/draw"
+)
+
+// Contour implements the Plotter interface, drawing
+// a contour plot of the values in the GridXYZ field. The
+// row index is from top to bottom.
+type Contour struct {
+	GridXYZ GridXYZ
+
+	// Levels describes the contour heights to plot.
+	Levels []float64
+
+	// LineStyle is the style of the contour lines.
+	draw.LineStyle
+
+	// Palette is the color palette used to render
+	// the heat map.
+	Palette palette.Palette
+
+	// Underflow and Overflow are colors used to draw
+	// contours outside the dynamic range defined
+	// by Min and Max.
+	Underflow color.Color
+	Overflow  color.Color
+
+	// Min and Max define the dynamic range of the
+	// heat map.
+	Min, Max float64
+}
+
+// NewContour creates as new contour plotter for the given data,
+// using the provided palette.
+func NewContour(g GridXYZ, levels []float64, p palette.Palette) *Contour {
+	var min, max float64
+	type minMaxer interface {
+		Min() float64
+		Max() float64
+	}
+	switch g := g.(type) {
+	case minMaxer:
+		min, max = g.Min(), g.Max()
+	default:
+		min, max = math.Inf(1), math.Inf(-1)
+		c, r := g.Dims()
+		for i := 0; i < c; i++ {
+			for j := 0; j < r; j++ {
+				v := g.Z(i, j)
+				if math.IsNaN(v) {
+					continue
+				}
+				min = math.Min(min, v)
+				max = math.Max(max, v)
+			}
+		}
+	}
+
+	if len(levels) == 0 {
+		levels = quantilesR7(g, defaultQuantiles)
+	}
+
+	return &Contour{
+		GridXYZ:   g,
+		Levels:    levels,
+		LineStyle: DefaultLineStyle,
+		Palette:   p,
+		Min:       min,
+		Max:       max,
+	}
+}
+
+// Default quantiles for case where levels is not explicitly set.
+var defaultQuantiles = []float64{0.01, 0.05, 0.25, 0.5, 0.75, 0.95, 0.99}
+
+// quantilesR7 returns the pth quantiles of the data in g according the th R-7 method.
+// http://en.wikipedia.org/wiki/Quantile#Estimating_the_quantiles_of_a_population
+func quantilesR7(g GridXYZ, p []float64) []float64 {
+	c, r := g.Dims()
+	data := make([]float64, 0, c*r)
+	for i := 0; i < c; i++ {
+		for j := 0; j < r; j++ {
+			data = append(data, g.Z(i, j))
+		}
+	}
+	sort.Float64s(data)
+	v := make([]float64, len(p))
+	for j, q := range p {
+		if q == 1 {
+			v[j] = data[len(data)-1]
+		}
+		h := float64(len(data)-1) * q
+		i := int(h)
+		v[j] = data[i] + (h-math.Floor(h))*(data[i+1]-data[i])
+	}
+	return v
+}
+
+// naive is a debugging constant. If true, Plot performs no contour path
+// reconstruction, instead rendering wach path segment individually.
+const naive = false
+
+// Plot implements the Plot method of the plot.Plotter interface.
+func (h *Contour) Plot(c draw.Canvas, plt *plot.Plot) {
+	var pal []color.Color
+	if h.Palette != nil {
+		pal = h.Palette.Colors()
+	}
+	if len(pal) == 0 {
+		pal = []color.Color{h.LineStyle.Color}
+	}
+
```

Comment:
I think the code eventually panics if len(pal) < len(h.Levels). There should be an explicit check / documentation